Code/6-wire-version/src/main.cpp

```diff
@@ -21,78 +21,144 @@ void setup() {
+  updateWSTimer.attach(2.0, []{ sendWSFlag = true; });
+
+  // update MQTT every 10 minutes. (will also be updated on state changes)
+  updateMqttTimer.attach(600, []{ sendMQTTFlag = true; });
+
+  // needs to be loaded here for reading the wifi.json
+  LittleFS.begin();
   loadWifi();
   startWiFi();
   startNTP();
   startOTA();
-  startServer();
+  startHttpServer();
   startWebSocket();
   startMQTT();
+
   bwc.print(WiFi.localIP().toString());
-  Serial.println(F("\nEnd of setup()"));
+  Serial.println(F("End of setup()"));
 }
 
-void loop() {
-  bwc.loop();                     // Fiddle with the pump computer
-  if(WiFi.status() == WL_CONNECTED){
-    webSocket.loop();             // check for websocket events
-    server.handleClient();        // run the server
-    ArduinoOTA.handle();          // listen for OTA events
-      
-    if (enableMqtt){
-      if(!MQTTclient.loop()){
+void loop()
+{
+  // Fiddle with the pump computer
+  bwc.loop();
+
+  // do things only when WiFi is connected
+  if (WiFi.status() == WL_CONNECTED)
+  {
+    // listen for websocket events
+    webSocket.loop();
+    // listen for webserver events
+    server.handleClient();
+    // listen for OTA events
+    ArduinoOTA.handle();
+    
+    // MQTT
+    if (enableMqtt)
+    {
+      if (!mqttClient.loop())
+      {
         MQTT_Connect();
-      } else {
+      }
+      else
+      {
         String msg = bwc.getButtonName();
-        //publish pretty button name if display button is pressed (or NOBTN if released)
-        if(!msg.equals(prevButtonName)) {
-          MQTTclient.publish((String(mqttBaseTopic) + "/button").c_str(), String(msg).c_str(), true);
+        // publish pretty button name if display button is pressed (or NOBTN if released)
+        if (!msg.equals(prevButtonName))
+        {
+          mqttClient.publish((String(mqttBaseTopic) + "/button").c_str(), String(msg).c_str(), true);
           prevButtonName = msg;
         }
-        if (bwc.newData()) sendMQTT();
+
+        if (bwc.newData())
+        {
+          sendMQTT();
+        }
+        else if (sendMQTTFlag)
+        {
+          sendMQTTFlag = false;
+          sendMQTT();
+        }
       }
     }
     
-    if (bwc.newData()) {
+    // web socket
+    if (bwc.newData())
+    {
       sendWS();
     }
-    if (sendWSFlag) {
+    else if (sendWSFlag)
+    {
       sendWSFlag = false;
       sendWS();
     }
-    if (sendMQTTFlag && enableMqtt) {
-      sendMQTTFlag = false;
-      sendMQTT();
+  }
+
+  // run when wifi connection lost
+  if (WiFi.status() != WL_CONNECTED)
+  {
+    // run once after connection was lost
+    if (wifiConnected)
+    {
+      Serial.println("WiFi > Lost connection. Trying to reconnect ...");
+    }
+    // set marker
+    wifiConnected = false;
+  }
+
+  // run when wifi connection established
+  if (WiFi.status() == WL_CONNECTED)
+  {
+    // run once after connection was established
+    if (!wifiConnected)
```

Comment:
Isn't the check for WL_CONNECTED redundant? Maybe insert this somewhere into the first test (line 54)?

Reply:
on the other hand, it does not do anything useful. I introduced the wifiConnected flag to restart wifi manually every 60 seconds, but I think it's obsolete in current code?

Reply:
i used it for printing the IP to the monitoring log, also showing that we got disconnected "now".

Reply:
i would keep it for debugging